src/databricks/sql/ae.py

```diff
@@ -0,0 +1,121 @@
+from enum import Enum
+from typing import Optional, Union, TYPE_CHECKING
+from databricks.sql.results import ResultSet
+
+
+if TYPE_CHECKING:
+    from databricks.sql.thrift_backend import ThriftBackend
+    from databricks.sql.client import Connection
+    from databricks.sql.results import ResultSet
+
+
+from uuid import UUID
+
+from databricks.sql.thrift_api.TCLIService import ttypes
+
+
+class AsyncExecutionStatus(Enum):
+    """An enum that represents the status of an async execution"""
+
+    PENDING = 0
+    RUNNING = 1
+    FINISHED = 2
+    CANCELED = 3
+    FETCHED = 4
+    ABORTED = 5
+
+
+def _toperationstate_to_ae_status(
+    input: ttypes.TOperationState,
+) -> AsyncExecutionStatus:
+    x = ttypes.TOperationState
+
+    if input in [x.INITIALIZED_STATE, x.PENDING_STATE, x.RUNNING_STATE]:
+        return AsyncExecutionStatus.RUNNING
+    if input == x.CANCELED_STATE:
+        return AsyncExecutionStatus.CANCELED
+    if input == x.FINISHED_STATE:
+        return AsyncExecutionStatus.FINISHED
+    if input in [x.CLOSED_STATE, x.ERROR_STATE, x.UKNOWN_STATE, x.TIMEDOUT_STATE]:
+        return AsyncExecutionStatus.ABORTED
+
+
+class AsyncExecution:
+    """
+    A class that represents an async execution of a query. Exposes just two methods:
+    get_result_or_status and cancel
+    """
+
+    _connection: "Connection"
+    _result_set: Optional["ResultSet"]
+    _execute_statement_response: Optional[ttypes.TExecuteStatementResp]
+
+    def __init__(
+        self,
+        connection: "Connection",
+        query_id: UUID,
+        status: AsyncExecutionStatus,
+        execute_statement_response: Optional[ttypes.TExecuteStatementResp] = None,
+    ):
+        self._connection = connection
+        self._execute_statement_response = execute_statement_response
+        self.query_id = query_id
+        self.status = status
+
+    status: AsyncExecutionStatus
+    query_id: UUID
+
+    def get_result_or_status(self) -> Union["ResultSet", AsyncExecutionStatus]:
+        """Get the result of the async execution. If execution has not completed, return False."""
+
+        if self.status == AsyncExecutionStatus.FINISHED:
+            self._thrift_fetch_result()
```

Comment:
Advisory: separate status checking from returning results.  It makes consumers' lives more complicated when a method/function can return two very different types.

Reply:
Works for me. The combination method was part of the original design but there's nothing sacred about it. In a follow-up PR I'll split this into two separate methods.

Reply:
Another option is to have a more structured type:

```
@dataclass
class ExecutionResult:
   status: AsyncExecutionStatus
   results: Optional[ResultSet]
```

Reply:
At least going with a dataclass like above means the consumer never has to match again type, only status.